pennylane_lightning/lightning_qubit/_adjoint_jacobian.py

```diff
@@ -0,0 +1,344 @@
+
+        if not tape_return_type:  # the tape does not have measurements
+            return np.array([], dtype=self._dtype)
+
+        if tape_return_type is State:
+            raise QuantumFunctionError(
+                "This method does not support statevector return type. "
+                "Use vjp method instead for this purpose."
+            )
+
+        self._check_adjdiff_supported_operations(tape.operations)
+
+        processed_data = self._process_jacobian_tape(tape)
+
+        if not processed_data:  # training_params is empty
+            return np.array([], dtype=self._dtype)
+
+        trainable_params = processed_data["tp_shift"]
+
+        # If requested batching over observables, chunk into OMP_NUM_THREADS sized chunks.
+        # This will allow use of Lightning with adjoint for large-qubit numbers AND large
+        # numbers of observables, enabling choice between compute time and memory use.
+        requested_threads = int(getenv("OMP_NUM_THREADS", "1"))
+
+        if self._batch_obs and requested_threads > 1:
+            obs_partitions = _chunk_iterable(processed_data["obs_serialized"], requested_threads)
+            jac = []
+            for obs_chunk in obs_partitions:
+                jac_local = self._jacobian_lightning(
+                    processed_data["state_vector"],
+                    obs_chunk,
+                    processed_data["ops_serialized"],
+                    trainable_params,
+                )
+                jac.extend(jac_local)
+        else:
+            jac = self._jacobian_lightning(
+                processed_data["state_vector"],
+                processed_data["obs_serialized"],
+                processed_data["ops_serialized"],
+                trainable_params,
+            )
+        jac = np.array(jac)
+        jac = jac.reshape(-1, len(trainable_params))
+        jac_r = np.zeros((jac.shape[0], processed_data["all_params"]))
+        jac_r[:, processed_data["record_tp_rows"]] = jac
+        if hasattr(qml, "active_return"):  # pragma: no cover
+            return self._adjoint_jacobian_processing(jac_r) if qml.active_return() else jac_r
+        return self._adjoint_jacobian_processing(jac_r)
+
+    # pylint: disable=inconsistent-return-statements
+    def calculate_vjp(self, tape: QuantumTape, grad_vec):
+        """Generate the processing function required to compute the vector-Jacobian products
+        of a tape.
+
+        .. code-block:: python
+
+            vjp_f = dev.vjp([qml.state()], grad_vec)
+            vjp = vjp_f(tape)
+
+        computes :math:`w = (w_1,\\cdots,w_m)` where
+
+        .. math::
+
+            w_k = \\langle v| \\frac{\\partial}{\\partial \\theta_k} | \\psi_{\\pmb{\\theta}} \\rangle.
+
+        Here, :math:`m` is the total number of trainable parameters, :math:`\\pmb{\\theta}`
+        is the vector of trainable parameters and :math:`\\psi_{\\pmb{\\theta}}`
+        is the output quantum state.
```

Comment:
What is $v$ here?

Reply:
I will come back to this later. I think the documentation needs some improvements here.